distributed/deploy/tests/test_adaptive.py

```diff
@@ -449,3 +449,40 @@ async def test_update_adaptive(cleanup):
         await asyncio.sleep(0.2)
         assert first.periodic_callback is None
         assert second.periodic_callback.is_running()
+
+
+def test_adaptive_too_many_workers():
```

Comment:
I'm not too familiar with `async/await` but I assume you would like to have tests with `sleep` to by `async`. I tried to async'ify this test but I get the following error:

```
_____________________________________________________________________________ ERROR at teardown of test_adaptive_too_many_workers ______________________________________________________________________________

    @pytest.fixture
    def cleanup():
        with clean():
>           yield

distributed/utils_test.py:1541: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../local/lib/python3.8/contextlib.py:120: in __exit__
    next(self.gen)
distributed/utils_test.py:1535: in clean
    del thread_state.on_event_loop_thread
../../local/lib/python3.8/contextlib.py:120: in __exit__
    next(self.gen)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    @contextmanager
    def check_thread_leak():
        active_threads_start = set(threading._active)
    
        yield
    
        start = time()
        while True:
            bad = [
                t
                for t, v in threading._active.items()
                if t not in active_threads_start
                and "Threaded" not in v.name
                and "watch message" not in v.name
                and "TCP-Executor" not in v.name
            ]
            if not bad:
                break
            else:
                sleep(0.01)
            if time() > start + 5:
                from distributed import profile
    
                tid = bad[0]
                thread = threading._active[tid]
                call_stacks = profile.call_stack(sys._current_frames()[tid])
>               assert False, (thread, call_stacks)
E               AssertionError: (<Thread(Profile, started daemon 140229604820736)>, ['  File "/results/twoertwe/home/local/lib/python3.8/threading.py"...', '  File "/results/twoertwe/home/git/distributed/distributed/profile.py", line 269, in _watch
E                 \tsleep(interval)
E                 '])
E               assert False

distributed/utils_test.py:1435: AssertionError
```
using the following code (not yet pushed):

```py
@pytest.mark.asyncio
async def test_adaptive_too_many_workers(cleanup):
    """
    Make sure that not more workers are spawned than runnable tasks.

    GH 4155
    """
    n = 4
    graph = {f"test_{i}": (slowinc, 1) for i in range(n)}

    def _assert_not_too_many_worker(dask_worker):
        assert dask_worker.name < n

    workers = 10  # we shouldn't need more than n=4 workers
    async with LocalCluster(
        n_workers=0,
        processes=True,
        threads_per_worker=1,
        memory_limit=0,
        asynchronous=True,
    ) as cluster:
        async with Client(cluster, asynchronous=True) as client:
            client.register_worker_callbacks(_assert_not_too_many_worker)
            cluster.adapt(minimum=1, maximum=workers)
            client.get(graph, list(graph.keys()))

```